### PR TITLE
#33387: fix GraphQl products query layered navigation category child

### LINKED ISSUE #33387: GraphQl products query layered navigation filters return incorrect child categories list
### Preconditions (*)

1. Magento 2.4.1 Adobe Commerce & Open Souce both.
2. Use sample data to reproduce the issue

### Steps to reproduce (*)

1. Use below GraphqlQl request with sample data setup 
Use category ID - 21 with default store code in the header

`
query {
  products(filter: {category_id: {eq: "21"}}) {
    aggregations {
      attribute_code
      count
      label
      options {
        count
        label
        value
      }
    }
  }
}
`

### Expected result (*)

1. Only these 4 sub-categories should appear in GraphQl response.

image

### Actual result (*)

1. More than 4 categories in aggregation result -

image

The response is more than 4 child categories which wrong here (same response below) -

`        {
          "attribute_code": "category_id",
          "count": 11,
          "label": "Category",
          "options": [
            {
              "count": 50,
              "label": "Tops",
              "value": "21"
            },
            {
              "count": 14,
              "label": "Bras & Tanks",
              "value": "26"
            },
            {
              "count": 13,
              "label": "New Luma Yoga Collection",
              "value": "8"
            },
            {
              "count": 13,
              "label": "Performance Fabrics",
              "value": "35"
            },
            {
              "count": 12,
              "label": "Jackets",
              "value": "23"
            },
            {
              "count": 12,
              "label": "Hoodies & Sweatshirts",
              "value": "24"
            },`

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by korovitskyi**
### Description (*)
It was investigated the category grid page -> category filters. There we check the category filter by current category ID and children. 
In the same way, was added a point to `\Magento\CatalogGraphQl\Model\Resolver\Products\Query\Search` to modify the aggregation category by searchable category ID. As for me, it is the best point, as we must know if request queries have been filtered by category_id and know ID value(s).
Reopen by https://github.com/magento/magento2/pull/33520.
### Fixed Issues (if relevant)

1. Fixes magento/magento2#33387

### Manual testing scenarios (*)
Send request and the count of categories should be 4:
```
query { 
  products(filter: {category_id: {eq: "21"}}) {
    aggregations {
      attribute_code
      count
      label
      options { 
        count
        label
        value 
      } 
    }
  } 
}
```

```
query { 
  products(filter: {category_id: {in: ["21"]}}) {
    aggregations {
      attribute_code
      count
      label
      options { 
        count
        label
        value 
      } 
    }
  } 
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    /**
+     * Get resolved category aggregation.
+     *
+     * @param array $categoryFilter
+     * @param Bucket[] $bucketList
+     *
+     * @return Aggregation
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Hello @korovitskyi 
Could you please merge tags to the common block?
Remove redundant line.

```suggestion
     * @param array $categoryFilter
     * @param Bucket[] $bucketList
     * @return Aggregation
```

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    {
+        $categoryBucket = $bucketList[self::CATEGORY_BUCKET] ?? [];
+        $values = $categoryBucket->getValues();
+        $condition = array_key_first($categoryFilter);
+        $searchableCategoryValue = $categoryFilter[$condition] ?? 0;
+
+        if (!$searchableCategoryValue || empty($values)) {
+            return $this->aggregationFactory->create([self::BUCKETS => $bucketList]);
+        }
+
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Please remove redundant line

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        $resolvedCategoryBucket = $this->bucketFactory->create(
+            [
+                'name' => self::CATEGORY_BUCKET,
+                'values' => $resolvedValues
+            ]
+        );
+
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Please remove redundant lines

```suggestion
        $resolvedCategoryBucket = $this->bucketFactory->create(
            [
                'name' => self::CATEGORY_BUCKET,
                'values' => $resolvedValues
            ]
        );
```

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * Check is valid searchable category children.
+     *
+     * @param int $searchableCategoryId
+     * @param AggregationValueInterface[] $aggregationValues
+     *
+     * @return AggregationValueInterface[]
+     *
+     * @throws NoSuchEntityException
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Hello @korovitskyi 
Could you please merge tags to the common block?
Remove redundant lines.

```suggestion
     * @param int $searchableCategoryId
     * @param AggregationValueInterface[] $aggregationValues
     * @return AggregationValueInterface[]
     * @throws NoSuchEntityException
```

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+     */
+    private function getValidCategories(int $searchableCategoryId, array $aggregationValues): array
+    {
+        $stareId = (int) $this->storeManager->getStore()->getId();
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Hello @korovitskyi 
Perhaps this is typo ($stareId should be named $storeId)

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+        foreach ($aggregationValues as $bucketValue) {
+            $childCategoryId = (int) $bucketValue->getValue();
+
+            if (!in_array($childCategoryId, $validChildIdList) ||
+                in_array($childCategoryId, $this->resolvedChildrenIds)
+            ) {
+                continue;
+            }
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Hello @korovitskyi 
Could you please fix it according to PSR-12 requirements and examples https://www.php-fig.org/psr/psr-12/ (see p.5.1)

```suggestion
            if (
                !in_array($childCategoryId, $validChildIdList)
                ||. in_array($childCategoryId, $this->resolvedChildrenIds)
            ) {
                continue;
            }
```

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            ]
+        );
+
+        $resolvedBucketList[self::CATEGORY_BUCKET] = $resolvedCategoryBucket;
+
+        return $this->aggregationFactory->create([self::BUCKETS => $resolvedBucketList]);
+    }
+
+    /**
+     * Check is valid searchable category children.
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php]: Hello @korovitskyi
This method not only checks, but returns list of categories, right?
This summary can confuse.
Could you please fix it?

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php [hunk 8]
```diff
@@ -84,6 +91,7 @@ public function __construct(
         $this->fieldSelection = $fieldSelection;
         $this->productsProvider = $productsProvider;
         $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->resolveCategoryAggregation = $resolveCategoryAggregation;
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php]: Hello @korovitskyi 
These changes are backward incompatible.
Could you please fix it.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php [hunk 9]
```diff
@@ -94,7 +102,9 @@ public function __construct(
      * @param array $args
      * @param ResolveInfo $info
      * @param ContextInterface $context
+     *
      * @return SearchResult
+     *
      * @throws GraphQlInputException
      */
```
Comment (andrewbess) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php]:
```suggestion
     * @param array $args
     * @param ResolveInfo $info
     * @param ContextInterface $context
     * @return SearchResult
     * @throws GraphQlInputException
     */
```

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @korovitskyi 
Thank you for your contribution.
Could you please add some fixes?
Thank you in advance

**Comment by korovitskyi** [comment]
Hi, @andrewbess .
Changes were added.
Thank you for the review.

**Comment by engcom-Charlie** [comment]
Hi @korovitskyi,

Thanks for your contribution!

As per your [comment](https://github.com/magento/magento2/pull/33547#issuecomment-886175666), you have implemented the review changes so moving it for review.

**Pull request merged**